keps/168-expose-pending-workloads-order/README.md

```diff
@@ -0,0 +1,259 @@
+# KEP-168: Expose order of pending workloads
+
+<!--
+This is the title of your KEP. Keep it short, simple, and descriptive. A good
+title can help communicate what the KEP is and should be considered as part of
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a KEP and for
+highlighting any additional information provided beyond the standard KEP
+template.
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+The enhancement extends the API of LocalQueue and ClusterQueue to expose the
+information about the order of their pending workloads.
+
+## Motivation
+
+Currently, there is no visibility to users of the contests of the queue by a
+user.
+.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- expose the order of workloads in the LocalQueue and ClusterQueue
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- expose the information about workload position for each workload individually
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to extend the APIs for the status of LocalQueue and ClusterQueue
+to expose the order to workloads. The order will be only exposed up to some
+configurable depth.
+
+The approach of keeping the information in the status of
+the queues will allow to expose the information without an extra cost of new
+API requests, as the statuses are already updated on changes for the number of
```

Comment:
That is not entirely true. Every time an object is updated, it triggers another reconcile. As such, if ordering is constantly changing, we will continue triggering reconciles. That is also true today, but the number of pending or active workloads might not change as frequently.

So we should aim to do the minimum number of updates possible. In particular, I think we don't need to update ClusterQueues. Order in the queue is something that end-users care about, but not so much cluster administrators.

Reply:
My idea was to just keep ordered lists, without positional information:
- LQ1: [a,b,c], pendingWorkloads: 3
- LQ2: [x,y], pendingWorkloads: 2
- Then the corresponding cluster queue could be [a,x,y,b,c]. 

IIUC in this case the change in the list will be pretty much equivalent to change in the `pendingWorkloads` counters.

Indeed, if we keep positional information inside localqueues:
- LQ1: `[a:0,b:3,c:4]`, pendingWorkloads: 3
- LQ2: `[x:1,y:2]`, pendingWorkloads: 2

Then, indeed, a change in LQ2 can impact the status of LQ1. Wondering if the benefits of getting the position from local queue are enough to complicate the design implementation with proper throttling.

Reply:
I see benefits of exposing the positional information inside local queues so I'm now leaning towards the mixed option: do not modify cluster queue, but have to configuration parameters:
- `exposeLocalQueueOrderDepth` - to control exposing the order which should not trigger extra requests
- `exposeClusterQueueIndexesDepth` - to control exposing the positional information which will require extra requests as updates to one local queue can impact order in another local queue

Reply:
Resolving as the design has changed. Now I've added the throttling with 5s intervals between updates to the structure. This should allow us to also include the positional information.